### PR TITLE
Cache ILM policy name on IndexMetadata

### LINKED ISSUE #83582: Make LIFECYCLE_NAME_SETTING a Field in IndexMetadata
Just like we did previously in e.g. https://github.com/elastic/elasticsearch/pull/78672 we can make the string in `LIFECYCLE_NAME_SETTING ` a field in `Indexmetadata` so we don't have to look it up for every index on every cluster state update in the lifecycle service.

This setting alone currently takes almost 1% of the overall master runtime (and more than that in terms of wall-time I would assume) when creating the initial indices set during many-shards benchmarking:

image

relates https://github.com/elastic/elasticsearch/issues/77466

### PR DESCRIPTION
**Pull request by joegallo**
Closes #83582

One little detail is that the `LIFECYCLE_NAME` value is defined on [IndexMetadata](https://github.com/elastic/elasticsearch/pull/83603/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R822) but it's still almost entirely referenced by way of [LifecycleSettings](https://github.com/elastic/elasticsearch/pull/83603/files#diff-4e8f046e86e7ec85e51de358c9773903205bebc9b34afacaab5d7d479837a39cR20). We need to have the value available in IndexMetadata in order to pull the value from the settings and set it into `lifecyclePolicyName`, but besides that everything about this still seems more `LifecycleSettings`-related to me.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java [hunk 1]
```diff
@@ -35,7 +34,7 @@
     // already mounted as a searchable snapshot. Those ILM actions will check if the index has this setting name configured.
     public static final String SNAPSHOT_INDEX_NAME = "index.store.snapshot.index_name";
 
-    public static final Setting<TimeValue> LIFECYCLE_POLL_INTERVAL_SETTING = timeSetting(
+    public static final Setting<TimeValue> LIFECYCLE_POLL_INTERVAL_SETTING = Setting.timeSetting(
```
Comment (original-brownbear) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:37]: This seems unrelated?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 2]
```diff
@@ -1642,6 +1656,7 @@ public IndexMetadata build() {
                 DiskThresholdDecider.SETTING_IGNORE_DISK_WATERMARKS.get(settings),
                 tierPreference,
                 ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.get(settings),
+                settings.get(IndexMetadata.LIFECYCLE_NAME),
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:1659]: This seems mildly hacky, should we just move the full Setting instance in here so we can just use `LIFECYCLE_NAME_SETTING.get(settings)` the standard way?

**Review by original-brownbear — COMMENTED** [review]
Thanks Joe! just one quick question.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java (hunk 2 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:1659]: This has the `null` semantics I want, but the standard way doesn't -- `SOME_SETTING.get(settings)` returns the default value for the setting if the setting is absent, and in this case that's an empty string (because you can't define the default value for a setting as being `null`).

Of course, that's not impossible to work around, we could do a `hasText` check or something like that.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:1659]: Ah I see. This is mildly scary because we seem to have been mixing both patterns for this setting before and now we're only using one that has `null` in it.
But I guess we use `if (Strings.isNullOrEmpty(policyName) == false) {` or the return as the right hand side of an equals so it should be good.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 3]
```diff
@@ -759,6 +765,10 @@ public int getShardsPerNodeLimit() {
         return tierPreference;
     }
 
+    public String getLifecyclePolicyName() {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:768]: Let's add `@Nullable` here

**Review by original-brownbear — APPROVED** [review]
LGTM (maybe add the annotation though, makes everyone's life easier). Thanks Joe!

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java (hunk 3 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:772]: 👍, https://github.com/elastic/elasticsearch/pull/83603/commits/508447c8efa1f77d8d0de4e3dac6054a9271b90b

**Comment by joegallo** [comment]
@elasticmachine run elasticsearch-ci/bwc
